### ISSUE TITLE
icosa_sphere vertex adjacencies?

### ISSUE THREAD

**Issue by MightyBOBcnc** [nschloe/meshzoo#74]
Hello,

Thank you for meshzoo. It's a very cool library.

For my project I am building an array of vertex adjacencies for an icosa_sphere and I was wondering if there is any structure or pattern with the order of vertex indices in the `cells` array in relation to a vertex's neighbors. (Aside from the 3 verts in each cell winding counter-clockwise, that is.)

In other words, given any vertex index on a sphere of any level of subdivision, could there be a simple function to get the indices of its 5 or 6 neighbors in either a consistent clockwise or consistent counter-clockwise manner?

For example say we provide the index of vertex 38 as in the below screenshot, could we get its neighbors (starting from any neighbor, let's say vertex 2) from meshzoo's cells array and output like `[2, 28, 12, 3, 23, 40]` or like `[2, 40, 23, 3, 12, 28]`?

icosa_sphere with n == 2
![image](https://user-images.githubusercontent.com/4467749/128740711-0c8e803f-2670-415f-aaad-40929228b4ee.png)

Thanks.

p.s. Have you ever considered adding numba `@jit` / `@njit` decorators in meshzoo?

**Comment by nschloe** [comment]
I don't think it's possible to infer that kind of structure from the meshzoo ordering, and if you can, it's an implementational detail that could change at any time.

In my experience, it's always worthwhile thinking about whether you _need_ this kind of information. The reason is that top-down data structures (cell -> edge or, edge->node, or cell -> node) are _much_ easier to handle than bottom-up structures (node -> cell, node -> neighbors etc.). One reason is that top-down structures are always well-shaped in memory (you know in advance that for 473 cells, you have a 473-by-3 array for cells->nodes). On the other hand, a node can be part of any number of cells and can have any number of neighbors. It's unclear how to efficiently store that in memory. Vectorized computation goes down the drain, too.

So, perhaps give it a moment to think about this.

> p.s. Have you ever considered adding numba @jit / @njit decorators in meshzoo?

Not really. I've found the speed-up we get by vectorizing everything more than compelling. Perhaps there are a few large Python loops left in meshzoo that could be optimized away, but in general I've never seen great need for jit.

**Comment by MightyBOBcnc** [comment]
Thank you for your thoughtful response.  

TL;DR up front:

* I've already written some very fast `build_adjacency` code to generate an "unordered" `vertex_neighbors` list for each vertex (I say "unordered" because it inherits from the orders in `cells` from `meshzoo.icosa_sphere`).  
* And I do have some `sort_adjacency` code to 'brute force' an ordered `vertex_neighbors` already but it's a bit slower than I'd like.  
* So my whole reason for asking is that I was hoping there was perhaps a pattern in the cell and vertex order returned by `cells` that I could use to shortcut the need for a brute-force sorting method.  
* However, if you don't think there is one then I suppose that's the end of this inquiry.  

In summation, all of the below explains the *why* of what I am doing.  

-----
 > In my experience, it's always worthwhile thinking about whether you *need* this kind of information.

I do have a need to know the neighbors for each vertex. The connectivity between vertices is basically a type of graph.  I'm attempting to implement a hydraulic erosion on a "sphere" and in order to do that I must compare the height of each vertex to its immediate neighbors.  Each vertex has its own height assigned from a Perlin/simplex noise. This is what it looks like after one octave of simplex height has been applied to the x,y,z coordinates of each vertex:
![image](https://user-images.githubusercontent.com/4467749/126717520-d8faf0a1-a502-48a0-bbb9-8f0c986c2e77.png)

A simplified explanation of the erosion would looks something like:
```python
for v in vertices:
    for n in vertex_neighbors:
        # compare height to v to determine a slope
        # move soil from higher verts to lower verts and adjust height accordingly
```

...where `vertices`, `vertex_neighbors`, and `heights` are all indexed to vertices and have the same length as `vertices`.

For this computation I don't need to know anything about the cells, only the connectivity to neighbor vertices.  Once I've built the `vertex_neighbors` adjacencies for each vertex the cells array becomes superfluous except for saving/visualizing the final result.

Technically speaking for the above pseudo-code example the order of neighbors in `vertex_neighbors` doesn't matter as all neighbors will be compared anyway.  

But I'm contemplating placing a restriction on the flow of water where having an ordered list of the neighbors might be more useful than an unordered one.  
Going back to the image from the first post, say a single rain drop is simulated as landing on vertex 2, and it is determined that it will flow to vertex 38.  If I don't want it to be able to turn back at an ugly acute angle to 28 or 40, and if I had an ordered neighbor list for `vertex_neighbors[38]` like `[2, 28, 12, 3, 23, 40]`, then invalid 'exit' verts would simply be the index of 2 (in this case 0), its index +1, and its index -1, or:
`vertex_neighbors[38][0]`
`vertex_neighbors[38][1]`
`vertex_neighbors[38][5]`

Maybe that's a stupid idea which will fail miserably but I'd rather experiment and see what happens than skip the idea entirely.  Maybe it will produce an interesting result, maybe it won't.

 > On the other hand, a node can be part of any number of cells and can have any number of neighbors.  

Au contraire, for an icosphere nodes 0 to 11 always have 5 neighbors and nodes 12 to N always have 6 so we do already know in advance how many neighbors each node has and how many cells it belongs to. (At least in the current implementation of meshzoo's `icosa_sphere`.)

Basic icosahedron (12 verts, 20 cells):
```
--vertex----vertex_neighbors----heights--
   0        [11 10  7  1  5]    [0.115] 
   1        [ 0  7  8  9  5]    [0.675] 
   2        [11  4  3  6 10]    [0.124] 
   3        [ 9  8  6  2  4]    [0.364] 
   4        [ 5  9  3  2 11]    [0.025] 
   5        [ 0  1  9  4 11]    [0.919] 
   6        [10  2  3  8  7]    [0.647] 
   7        [ 0 10  6  8  1]    [0.715] 
   8        [ 7  6  3  9  1]    [0.343] 
   9        [ 1  8  3  4  5]    [0.841] 
   10       [ 0 11  2  6  7]    [0.715] 
   11       [ 5  4  2 10  0]    [0.282] 
```
There are other reasons why it might be useful to have a short list of all neighbor verts for a given vertex.  A graph of edges can be inferred and reconstructed from the index and the known neighbors.
```
0  (0, 11) (0,10) (0,7) (0,1) (0, 5)
```
Or from the neighbors only.
```
0  (11,10) (10,7) (7,1) (1,5) (5,11)
```

It's possible that I will end up switching data structures later (maybe to a Delaunay triangulation of the vertices, or some sort of KD-Tree) and this is a fool's errand but for now this is the data structure I want to investigate.

**Comment by nschloe** [comment]
```python
for v in vertices:
    for n in vertex_neighbors:
        # compare height to v to determine a slope
        # move soil from higher verts to lower verts and adjust height accordingly
```
You can do this without computing the neighbors. For every node, you _aggregate_ the contributions from all of its neighbors, and for this, you only need cell connectivity.
```python
for c in cells:
    for i, j in neighbor_pairs:
        aggregates[i] += elevation[j]
        aggregates[j] += elevation[i]
```
In the end, you divide by 2 because you counted each contribution twice. Of course, you wouldn't use explicit Python loops, but `np.add.at` or [`npx.sum_at`](https://github.com/nschloe/npx#sum_atadd_at).

**Comment by MightyBOBcnc** [comment]
I'll look into a cells-based approach. 

As this thread isn't on-topic for meshzoo issues I'll go ahead and close this it now.  Thank you again for your time. :-)
